T/Torch/build_tarballs.jl

```diff
@@ -101,31 +111,20 @@ if [[ $target == x86_64-apple-darwin* # Fails to compile: /workspace/srcdir/pyto
     cmake_extra_args+="-DUSE_MKLDNN=OFF "
 fi
 
-if [[ $bb_full_target == *cuda* ]]; then
-    cuda_version=`echo $bb_full_target | sed -E -e 's/.*cuda\+([0-9]+\.[0-9]+).*/\1/'`
-    cuda_version_major=`echo $cuda_version | cut -d . -f 1`
-    cuda_version_minor=`echo $cuda_version | cut -d . -f 2`
-    cuda_full_path="$WORKSPACE/srcdir/CUDA_full.v$cuda_version/cuda"
-    export PATH=$PATH:$cuda_full_path/bin
-    export CUDACXX=$cuda_full_path/bin/nvcc
-    export CUDAHOSTCXX=$CXX
+cuda_version=${bb_full_target##*-cuda+}
+if [[ $bb_full_target == *cuda* ]] && [[ $cuda_version != none ]]; then
+    export CUDA_PATH="$prefix/cuda"
     mkdir $WORKSPACE/tmpdir
     export TMPDIR=$WORKSPACE/tmpdir
     cmake_extra_args+="\
+        -DCMAKE_CUDA_COMPILER_LAUNCHER=ccache \
```

Comment:
Just checking if ccache can manage the CUDA/nvcc calls.

Reply:
Ccache documentation says they support nvcc

Reply:
Yeah, it's not obvious though from the build logs that its having any effect...

It would be nice to be able to see the `ccache --show-stats`...

But it may also be something for later (again): https://github.com/JuliaPackaging/Yggdrasil/pull/4554#discussion_r950682841

Reply:
😀 

> Does adding CMAKE_CUDA_COMPILER_LAUNCHER actually help?

https://github.com/pytorch/pytorch/issues/8433#issuecomment-571119074

Reply:
Ah - that was superfluous - the PyTorch CMake set-up already handles setting `CMAKE_CUDA_COMPILER_LAUNCHER`: https://github.com/pytorch/pytorch/blob/v1.10.2/CMakeLists.txt#L331

And if one checks the logs, it is activated for both `x86_64-linux-gnu-cxx11`(`-cuda+none`), and, e.g. `x86_64-linux-gnu-cxx11-cuda+10.2`.

... and it appears in the log in the calls to `g++`,
```
cd /workspace/srcdir/pytorch/build/caffe2 && /usr/bin/ccache /opt/bin/x86_64-linux-gnu-libgfortran5-cxx11-cuda+10.2/x86_64-linux-gnu-g++ # ...
```
but not in the calls to `nvcc`,
```
[09:03:14] [ 81%] Building NVCC (Device) object caffe2/CMakeFiles/torch_cuda.dir/utils/torch_cuda_generated_math_gpu.cu.o
[09:03:14] cd /workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils && /usr/bin/cmake -D verbose:BOOL=1 -D build_configuration:STRING=Release -D generated_file:STRING=/workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/./torch_cuda_generated_math_gpu.cu.o -D generated_cubin_file:STRING=/workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/./torch_cuda_generated_math_gpu.cu.o.cubin.txt -P /workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/torch_cuda_generated_math_gpu.cu.o.Release.cmake
[09:03:14] -- Removing /workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/./torch_cuda_generated_math_gpu.cu.o
[09:03:14] /usr/bin/cmake -E remove /workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/./torch_cuda_generated_math_gpu.cu.o
[09:03:14] -- Generating dependency file: /workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/torch_cuda_generated_math_gpu.cu.o.NVCC-depend
[09:03:14] /workspace/destdir/cuda/bin/nvcc -M -D__CUDACC__ /workspace/srcdir/pytorch/caffe2/utils/math_gpu.cu -o /workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/torch_cuda_generated_math_gpu.cu.o.NVCC-depend -ccbin /opt/bin/x86_64-linux-gnu-libgfortran5-cxx11-cuda+10.2/x86_64-linux-gnu-gcc -m64 -Dtorch_cuda_EXPORTS -DUSE_CUDA -DTORCH_CUDA_BUILD_MAIN_LIB -DONNX_ML=1 -DONNXIFI_ENABLE_EXT=1 -DONNX_NAMESPACE=onnx_torch -DMAGMA_V2 -DHAVE_MMAP=1 -D_FILE_OFFSET_BITS=64 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DHAVE_MALLOC_USABLE_SIZE=1 -DUSE_EXTERNAL_MZCRC -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DUSE_DISTRIBUTED -DUSE_C10D_GLOO -DUSE_RPC -DUSE_TENSORPIPE -Xcompiler ,\"-D_GLIBCXX_USE_CXX11_ABI=1\",\"-Wno-deprecated\",\"-fvisibility-inlines-hidden\",\"-DUSE_PTHREADPOOL\",\"-fopenmp\",\"-DNDEBUG\",\"-DUSE_KINETO\",\"-DUSE_FBGEMM\",\"-DUSE_PYTORCH_QNNPACK\",\"-DUSE_XNNPACK\",\"-DSYMBOLICATE_MOBILE_DEBUG_HANDLE\",\"-DEDGE_PROFILER_USE_KINETO\",\"-O2\",\"-fPIC\",\"-Wno-narrowing\",\"-Wall\",\"-Wextra\",\"-Werror=return-type\",\"-Wno-missing-field-initializers\",\"-Wno-type-limits\",\"-Wno-array-bounds\",\"-Wno-unknown-pragmas\",\"-Wno-sign-compare\",\"-Wno-unused-parameter\",\"-Wno-unused-variable\",\"-Wno-unused-function\",\"-Wno-unused-result\",\"-Wno-unused-local-typedefs\",\"-Wno-strict-overflow\",\"-Wno-strict-aliasing\",\"-Wno-error=deprecated-declarations\",\"-Wno-stringop-overflow\",\"-Wno-psabi\",\"-Wno-error=pedantic\",\"-Wno-error=redundant-decls\",\"-Wno-error=old-style-cast\",\"-fdiagnostics-color=always\",\"-faligned-new\",\"-Wno-unused-but-set-variable\",\"-Wno-maybe-uninitialized\",\"-fno-math-errno\",\"-fno-trapping-math\",\"-Werror=format\",\"-Werror=cast-function-type\",\"-Wno-stringop-overflow\",\"-DHAVE_AVX2_CPU_DEFINITION\",\"-fPIC\",\"-O3\",\"-DNDEBUG\",\"-DNDEBUG\" -Xfatbin -compress-all -DONNX_NAMESPACE=onnx_torch -gencode arch=compute_35,code=sm_35 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_52,code=sm_52 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_70,code=sm_70 -gencode arch=compute_75,code=sm_75 -gencode arch=compute_75,code=compute_75 -Xcudafe --diag_suppress=cc_clobber_ignored,--diag_suppress=integer_sign_change,--diag_suppress=useless_using_declaration,--diag_suppress=set_but_not_used,--diag_suppress=field_without_dll_interface,--diag_suppress=base_class_has_different_dll_interface,--diag_suppress=dll_interface_conflict_none_assumed,--diag_suppress=dll_interface_conflict_dllexport_assumed,--diag_suppress=implicit_return_from_non_void_function,--diag_suppress=unsigned_compare_with_zero,--diag_suppress=declared_but_not_referenced,--diag_suppress=bad_friend_decl -std=c++14 -Xcompiler -fPIC --expt-relaxed-constexpr --expt-extended-lambda -Wno-deprecated-gpu-targets --expt-extended-lambda -Xcompiler -fPIC -DCUDA_HAS_FP16=1 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ -DNVCC -I/workspace/destdir/cuda/include -I/workspace/srcdir/pytorch/build/aten/src -I/workspace/srcdir/pytorch/aten/src -I/workspace/srcdir/pytorch/build -I/workspace/srcdir/pytorch -I/workspace/srcdir/pytorch/build/third_party/gloo -I/workspace/srcdir/pytorch/cmake/../third_party/gloo -I/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/local/include -I/workspace/srcdir/pytorch/third_party -I/workspace/srcdir/pytorch/cmake/../third_party/eigen -I/workspace/srcdir/pytorch/cmake/../third_party/pybind11/include -I/workspace/srcdir/pytorch/cmake/../third_party/cudnn_frontend/include -I/workspace/srcdir/pytorch/third_party/cub -I/workspace/srcdir/pytorch/build/caffe2/contrib/aten -I/workspace/srcdir/pytorch/third_party/onnx -I/workspace/srcdir/pytorch/build/third_party/onnx -I/workspace/srcdir/pytorch/third_party/foxi -I/workspace/srcdir/pytorch/build/third_party/foxi -I/root/micromamba/include -I/workspace/srcdir/pytorch/third_party/ideep/mkl-dnn/include -I/workspace/srcdir/pytorch/third_party/ideep/include -I/workspace/srcdir/pytorch/build/include -I/workspace/srcdir/pytorch/torch/csrc/distributed -I/workspace/srcdir/pytorch/build/caffe2/aten/src/TH -I/workspace/srcdir/pytorch/aten/src/TH -I/workspace/srcdir/pytorch/build/caffe2/aten/src/THC -I/workspace/srcdir/pytorch/aten/src/THC -I/workspace/srcdir/pytorch/aten/src/ATen/cuda -I/workspace/srcdir/pytorch/build/caffe2/aten/src -I/workspace/srcdir/pytorch/aten/../third_party/catch/single_include -I/workspace/srcdir/pytorch/aten/src/ATen/.. -I/workspace/srcdir/pytorch/build/caffe2/aten/src/ATen -I/workspace/srcdir/pytorch/c10/cuda/../.. -I/workspace/srcdir/pytorch/c10/../ -I/workspace/srcdir/pytorch/third_party/tensorpipe -I/workspace/srcdir/pytorch/build/third_party/tensorpipe -I/workspace/srcdir/pytorch/third_party/tensorpipe/third_party/libnop/include -I/workspace/srcdir/pytorch/torch/csrc/api -I/workspace/srcdir/pytorch/torch/csrc/api/include -I/workspace/srcdir/pytorch/build/third_party/ideep/mkl-dnn/include -I/workspace/srcdir/pytorch/third_party/ideep/mkl-dnn/src/../include
[09:03:27] -- Generating temporary cmake readable file: /workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/torch_cuda_generated_math_gpu.cu.o.depend.tmp
[09:03:27] /usr/bin/cmake -D input_file:FILEPATH=/workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/torch_cuda_generated_math_gpu.cu.o.NVCC-depend -D output_file:FILEPATH=/workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/torch_cuda_generated_math_gpu.cu.o.depend.tmp -D verbose=1 -P /workspace/srcdir/pytorch/cmake/Modules_CUDA_fix/upstream/FindCUDA/make2cmake.cmake
[09:03:28] -- Copy if different /workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/torch_cuda_generated_math_gpu.cu.o.depend.tmp to /workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/torch_cuda_generated_math_gpu.cu.o.depend
[09:03:28] /usr/bin/cmake -E copy_if_different /workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/torch_cuda_generated_math_gpu.cu.o.depend.tmp /workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/torch_cuda_generated_math_gpu.cu.o.depend
[09:03:28] -- Removing /workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/torch_cuda_generated_math_gpu.cu.o.depend.tmp and /workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/torch_cuda_generated_math_gpu.cu.o.NVCC-depend
[09:03:28] /usr/bin/cmake -E remove /workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/torch_cuda_generated_math_gpu.cu.o.depend.tmp /workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/torch_cuda_generated_math_gpu.cu.o.NVCC-depend
[09:03:28] -- Generating /workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/./torch_cuda_generated_math_gpu.cu.o
[09:03:28] /workspace/destdir/cuda/bin/nvcc /workspace/srcdir/pytorch/caffe2/utils/math_gpu.cu -c -o /workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/./torch_cuda_generated_math_gpu.cu.o -ccbin /opt/bin/x86_64-linux-gnu-libgfortran5-cxx11-cuda+10.2/x86_64-linux-gnu-gcc -m64 -Dtorch_cuda_EXPORTS -DUSE_CUDA -DTORCH_CUDA_BUILD_MAIN_LIB -DONNX_ML=1 -DONNXIFI_ENABLE_EXT=1 -DONNX_NAMESPACE=onnx_torch -DMAGMA_V2 -DHAVE_MMAP=1 -D_FILE_OFFSET_BITS=64 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DHAVE_MALLOC_USABLE_SIZE=1 -DUSE_EXTERNAL_MZCRC -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DUSE_DISTRIBUTED -DUSE_C10D_GLOO -DUSE_RPC -DUSE_TENSORPIPE -Xcompiler ,\"-D_GLIBCXX_USE_CXX11_ABI=1\",\"-Wno-deprecated\",\"-fvisibility-inlines-hidden\",\"-DUSE_PTHREADPOOL\",\"-fopenmp\",\"-DNDEBUG\",\"-DUSE_KINETO\",\"-DUSE_FBGEMM\",\"-DUSE_PYTORCH_QNNPACK\",\"-DUSE_XNNPACK\",\"-DSYMBOLICATE_MOBILE_DEBUG_HANDLE\",\"-DEDGE_PROFILER_USE_KINETO\",\"-O2\",\"-fPIC\",\"-Wno-narrowing\",\"-Wall\",\"-Wextra\",\"-Werror=return-type\",\"-Wno-missing-field-initializers\",\"-Wno-type-limits\",\"-Wno-array-bounds\",\"-Wno-unknown-pragmas\",\"-Wno-sign-compare\",\"-Wno-unused-parameter\",\"-Wno-unused-variable\",\"-Wno-unused-function\",\"-Wno-unused-result\",\"-Wno-unused-local-typedefs\",\"-Wno-strict-overflow\",\"-Wno-strict-aliasing\",\"-Wno-error=deprecated-declarations\",\"-Wno-stringop-overflow\",\"-Wno-psabi\",\"-Wno-error=pedantic\",\"-Wno-error=redundant-decls\",\"-Wno-error=old-style-cast\",\"-fdiagnostics-color=always\",\"-faligned-new\",\"-Wno-unused-but-set-variable\",\"-Wno-maybe-uninitialized\",\"-fno-math-errno\",\"-fno-trapping-math\",\"-Werror=format\",\"-Werror=cast-function-type\",\"-Wno-stringop-overflow\",\"-DHAVE_AVX2_CPU_DEFINITION\",\"-fPIC\",\"-O3\",\"-DNDEBUG\",\"-DNDEBUG\" -Xfatbin -compress-all -DONNX_NAMESPACE=onnx_torch -gencode arch=compute_35,code=sm_35 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_52,code=sm_52 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_70,code=sm_70 -gencode arch=compute_75,code=sm_75 -gencode arch=compute_75,code=compute_75 -Xcudafe --diag_suppress=cc_clobber_ignored,--diag_suppress=integer_sign_change,--diag_suppress=useless_using_declaration,--diag_suppress=set_but_not_used,--diag_suppress=field_without_dll_interface,--diag_suppress=base_class_has_different_dll_interface,--diag_suppress=dll_interface_conflict_none_assumed,--diag_suppress=dll_interface_conflict_dllexport_assumed,--diag_suppress=implicit_return_from_non_void_function,--diag_suppress=unsigned_compare_with_zero,--diag_suppress=declared_but_not_referenced,--diag_suppress=bad_friend_decl -std=c++14 -Xcompiler -fPIC --expt-relaxed-constexpr --expt-extended-lambda -Wno-deprecated-gpu-targets --expt-extended-lambda -Xcompiler -fPIC -DCUDA_HAS_FP16=1 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ -DNVCC -I/workspace/destdir/cuda/include -I/workspace/srcdir/pytorch/build/aten/src -I/workspace/srcdir/pytorch/aten/src -I/workspace/srcdir/pytorch/build -I/workspace/srcdir/pytorch -I/workspace/srcdir/pytorch/build/third_party/gloo -I/workspace/srcdir/pytorch/cmake/../third_party/gloo -I/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/local/include -I/workspace/srcdir/pytorch/third_party -I/workspace/srcdir/pytorch/cmake/../third_party/eigen -I/workspace/srcdir/pytorch/cmake/../third_party/pybind11/include -I/workspace/srcdir/pytorch/cmake/../third_party/cudnn_frontend/include -I/workspace/srcdir/pytorch/third_party/cub -I/workspace/srcdir/pytorch/build/caffe2/contrib/aten -I/workspace/srcdir/pytorch/third_party/onnx -I/workspace/srcdir/pytorch/build/third_party/onnx -I/workspace/srcdir/pytorch/third_party/foxi -I/workspace/srcdir/pytorch/build/third_party/foxi -I/root/micromamba/include -I/workspace/srcdir/pytorch/third_party/ideep/mkl-dnn/include -I/workspace/srcdir/pytorch/third_party/ideep/include -I/workspace/srcdir/pytorch/build/include -I/workspace/srcdir/pytorch/torch/csrc/distributed -I/workspace/srcdir/pytorch/build/caffe2/aten/src/TH -I/workspace/srcdir/pytorch/aten/src/TH -I/workspace/srcdir/pytorch/build/caffe2/aten/src/THC -I/workspace/srcdir/pytorch/aten/src/THC -I/workspace/srcdir/pytorch/aten/src/ATen/cuda -I/workspace/srcdir/pytorch/build/caffe2/aten/src -I/workspace/srcdir/pytorch/aten/../third_party/catch/single_include -I/workspace/srcdir/pytorch/aten/src/ATen/.. -I/workspace/srcdir/pytorch/build/caffe2/aten/src/ATen -I/workspace/srcdir/pytorch/c10/cuda/../.. -I/workspace/srcdir/pytorch/c10/../ -I/workspace/srcdir/pytorch/third_party/tensorpipe -I/workspace/srcdir/pytorch/build/third_party/tensorpipe -I/workspace/srcdir/pytorch/third_party/tensorpipe/third_party/libnop/include -I/workspace/srcdir/pytorch/torch/csrc/api -I/workspace/srcdir/pytorch/torch/csrc/api/include -I/workspace/srcdir/pytorch/build/third_party/ideep/mkl-dnn/include -I/workspace/srcdir/pytorch/third_party/ideep/mkl-dnn/src/../include
[09:06:22] Generated /workspace/srcdir/pytorch/build/caffe2/CMakeFiles/torch_cuda.dir/utils/./torch_cuda_generated_math_gpu.cu.o successfully.
```
Sigh.

Reply:
... though it might be possible to get ccache in via the `ccbin` arg to nvcc, IIRC - which is currently getting set to GCC `-ccbin /opt/bin/x86_64-linux-gnu-libgfortran5-cxx11-cuda+10.2/x86_64-linux-gnu-gcc`.

Reply:
But giving up on this for the time being - as in #4554 - as it seems it should be handled automatically by the PyTorch CMake set-up.